pageserver/src/tenant/mgr.rs

```diff
@@ -532,10 +540,15 @@ where
         Err(e) => {
             let tenants_accessor = TENANTS.read().await;
             match tenants_accessor.get(&tenant_id) {
-                Some(tenant) => tenant.set_broken(&e.to_string()),
-                None => warn!("Tenant {tenant_id} got removed from memory"),
+                Some(tenant) => {
+                    tenant.set_broken(&e.to_string());
+                    Err(TenantStateError::Broken(tenant_id))
+                }
+                None => {
+                    warn!("Tenant {tenant_id} got removed from memory");
+                    Err(TenantStateError::NotFound(tenant_id))
+                }
             }
-            Err(e)
         }
```

Comment:
Replying to my comment at https://github.com/neondatabase/neon/pull/3818/files#r1136739345:

Oh, so this is where the "other" error is swallowed. The `tenant_cleanup` function returns anyhow::Error, and it is turned into a TenantStateError here.

We don't propagate the error details to the caller anymore. That might make debugging harder.

So we check if the tenant exists after this call, and if it does not, we assume that the anyhow:Error was caused by that. That seems fragile. Another thread might've detached the tenant concurrently, for starters, and we might incorrectly conclude that our detach operation failed because the tenant didn't exist, even though the error was something different. We would leave no trace of the original error in the logs in that case.

If we set the tenant Broken,we pass `e.to_string()` to `set_broken`, which prints the error message to the log. Does `e.to_string()` include the stack trace and all?

Perhaps `set_broken` should take the full Error as argument, instead of just a String.

Reply:
I added back TenantStateError::Other for this case to allow the error to propagate 
e.to_string() has no stack trace, we can optionally call e.source() to get the orig error if available